File: android/app/src/main/java/com/clipsync/android/media/ImageChunks.kt
package com.clipsync.android.media

import java.util.Base64

data class ImageChunk(
    val index: Int,
    val count: Int,
    val byteCount: Int,
    val data: String,
)

object ImageChunks {
    fun split(encoded: ByteArray): List<ImageChunk> {
        require(encoded.isNotEmpty() && encoded.size <= MediaLimits.MAX_ENCODED_BYTES) {
            "Encoded image size is out of bounds."
        }
        val count = (encoded.size + MediaLimits.MAX_CHUNK_BYTES - 1) / MediaLimits.MAX_CHUNK_BYTES
        require(count <= MediaLimits.MAX_CHUNK_COUNT) { "Encoded image needs too many chunks." }
        return (0 until count).map { index ->
            val start = index * MediaLimits.MAX_CHUNK_BYTES
            val length = minOf(MediaLimits.MAX_CHUNK_BYTES, encoded.size - start)
            val slice = encoded.copyOfRange(start, start + length)
            ImageChunk(
                index = index,
                count = count,
                byteCount = length,
                data = encodeBase64Url(slice),
            )
        }
    }

    fun tryDecodeChunk(data: String, expectedBytes: Int): ByteArray? {
        val decoded = try {
            decodeBase64Url(data)
        } catch (_: IllegalArgumentException) {
            return null
        }
        if (decoded.size != expectedBytes || expectedBytes !in 1..MediaLimits.MAX_CHUNK_BYTES) {
            return null
        }
        return decoded
    }

    fun encodeBase64Url(bytes: ByteArray): String =
        Base64.getUrlEncoder().withoutPadding().encodeToString(bytes)

    fun decodeBase64Url(value: String): ByteArray =
        Base64.getUrlDecoder().decode(value)
}

File: windows/ClipSync.Core/Media/ImageChunks.cs
using ClipSync.Core.Protocol;

namespace ClipSync.Core.Media;

public sealed record ImageChunk(int Index, int Count, int ByteCount, string Data);

public static class ImageChunks
{
    public static IReadOnlyList<ImageChunk> Split(ReadOnlySpan<byte> encoded)
    {
        if (encoded.Length is < 1 or > MediaLimits.MaxEncodedBytes)
        {
            throw new ArgumentOutOfRangeException(nameof(encoded), "Encoded image size is out of bounds.");
        }

        var count = (encoded.Length + MediaLimits.MaxChunkBytes - 1) / MediaLimits.MaxChunkBytes;
        if (count > MediaLimits.MaxChunkCount)
        {
            throw new ArgumentOutOfRangeException(nameof(encoded), "Encoded image needs too many chunks.");
        }

        var chunks = new ImageChunk[count];
        for (var index = 0; index < count; index++)
        {
            var start = index * MediaLimits.MaxChunkBytes;
            var length = Math.Min(MediaLimits.MaxChunkBytes, encoded.Length - start);
            chunks[index] = new ImageChunk(
                index,
                count,
                length,
                ProtocolValidation.EncodeBase64Url(encoded.Slice(start, length)));
        }

        return chunks;
    }

    public static bool TryDecodeChunk(string data, int expectedBytes, out byte[] bytes)
    {
        bytes = [];
        if (!ProtocolValidation.TryDecodeBase64Url(data, out var decoded))
        {
            return false;
        }

        if (decoded.Length != expectedBytes
            || expectedBytes is < 1 or > MediaLimits.MaxChunkBytes)
        {
            return false;
        }

        bytes = decoded;
        return true;
    }
}
